### PR TITLE
add buttons to view and download captured html for each page snapshot

### PR DESCRIPTION
**Pull request by tommeagher**
Addresses #330.

Also adds DEVELOPING.md as a guide to getting Klaxon running on your local machine.

I have this working locally, though I'm sure my neophyte Rails knowledge makes this clumsier than it ought to be. Would love any and all advice. For instance, should the html and download methods be on the pages controller instead of the changes controller? 

Also, not really sure what, if anything, to do about the CSS of the captured html bleeding through on the `/html` preview pages, or if the changes.scss is necessary. Might just leave the html as is with the warning, but could probably test with a wider array of real sites.

### CODE REVIEW

File: DEVELOPING.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+After you've cloned it, `cd` into the Klaxon directory. If you're on a Mac, you'll need to already have [homebrew](https://brew.sh/) and [rbenv](https://github.com/rbenv/rbenv) (a program that manages versions of Ruby) installed. Then you'll want 
+
+These next commands will put the proper (albeit aged) version of Ruby that Klaxon requires on your machine and make it available in this repo's directory.
+
+```
+brew update && brew upgrade ruby-build
+rbenv install
+``` 
+
+Next, you'll need to have the proper versions of all of Klaxon's dependency libraries, as well as foreman, Heroku's tool for running a local server for development. Run these commands here.
```
Comment (GabeIsman) [inline DEVELOPING.md:20]: Minor detail, but foreman is not made by Heroku and is more general than this description would indicate. I would describe it as a "process manager".

File: DEVELOPING.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+In order to be allowed to login to Klaxon once it's running on your machine, it'll need to know that your email address is. Create a file named `.env` and paste in these two items: 
+
+```
+ADMIN_EMAILS="my_awesome_email@gmail.com"
+HOST='localhost:5000'
+```
+
+Feel free to substitute in your real email address.
```
Comment (GabeIsman) [inline DEVELOPING.md:35]: Worth noting perhaps that Klaxon doesn't actually send emails locally, so a real email is not required.

File: DEVELOPING.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
+# Developing Klaxon Locally #
+
+If you're interested in [helping improve Klaxon by contributing code](CONTRIBUTING.md), this is a quick guide to how to get it running on your local machine for development.
+
+First, we'll assume you already have git on your local machine, as well as a standard postgres installation. So go to your terminal, navigate to whatever directory where you keep your projects and clone the Klaxon repo.
```
Comment (GabeIsman) [inline DEVELOPING.md:5]: Might be worth mentioning and linking to https://postgresapp.com/ as an easy way to get Postgres up and running.

File: DEVELOPING.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+```
+rake db:create
+rake db:migrate
+```
+
+Now, you should be about ready to get started. This command in the top folder of the Klaxon repo will get your dev server rolling.
+
+```
+foreman start -f Procfile
```
Comment (GabeIsman) [inline DEVELOPING.md:47]: This works but the simpler `foreman start` should do the same thing.

File: DEVELOPING.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+Now, you should be able to go to `localhost:5000` in your web browser and see Klaxon's welcome screen pop up. You'll want to manually add a webpage or two to watch at ```watching/pages/new```. For development purposes, you'll probably want to pick a site that updates pretty regularly. We use `http://www.timeanddate.com/`.
```
Comment (GabeIsman) [inline DEVELOPING.md:50]: These urls could be links!

File: app/assets/stylesheets/changes.scss [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// Place all the styles related to the Changes controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
+
+iframe {
+	height: 800px;
+	width: 100%;
+}
```
Comment (GabeIsman) [inline app/assets/stylesheets/changes.scss:8]: I think these styles should just be inline in the html. This css will be general to all iframes in the codebase (not that there are others, but it lays a trap for future developers).

File: app/controllers/changes_controller.rb [hunk 7]
```diff
@@ -23,6 +23,34 @@ def resend
     render json: @change.send_notifications
   end
 
+  def html
+    @change_id=params[:change_id]
```
Comment (GabeIsman) [inline app/controllers/changes_controller.rb:27]: So instead of directly interacting with the params here, I would add these methods to the list using the `set_change` filter on line 2. That will populate the `@change` variable before each method is called.

File: app/controllers/changes_controller.rb [hunk 8]
```diff
@@ -23,6 +23,34 @@ def resend
...
+    @snapshot = PageSnapshot.find(@change_id)
```
Comment (GabeIsman) [inline app/controllers/changes_controller.rb:28]: I think this is faulty logic that looks like works because there happen to be the same number of changes as snapshots (all belonging to the same page) in your local database. The `change_id` refers to the `changes_table` but you're using to query the `page_snapshots` table. The correct way to get the relevant snapshot would be with `@change.after` (if you've followed my suggestion above, otherwise `Change.find(params[:change_id]).after`).

File: app/controllers/changes_controller.rb [hunk 9]
```diff
@@ -23,6 +23,34 @@ def resend
...
+    @page = Page.find(@snapshot.page_id)
```
Comment (GabeIsman) [inline app/controllers/changes_controller.rb:29]: Similarly this can be simplified using ActiveRecord relations to `@snapshot.page`. Also just as a general note if a piece of data is a simple relation, you don't really need to assign it to a controller variable, you can just reference the relation in the view. This can get you into trouble with complex or expensive relational queries, but in general Rails does a good job of smart caching and you shouldn't worry too much about it.

File: app/controllers/changes_controller.rb [hunk 10]
```diff
@@ -23,6 +23,34 @@ def resend
...
+    @timestamp = @snapshot.created_at
+    @html = @snapshot.html
```
Comment (GabeIsman) [inline app/controllers/changes_controller.rb:31]: Similarly I wouldn't assign these pieces of data to their own variables, it just obfuscates where the data is coming from to someone looking at the view.

File: app/controllers/changes_controller.rb [hunk 11]
```diff
@@ -23,6 +23,34 @@ def resend
...
+    @change_id = @id
+    @snapshot = PageSnapshot.find(@id)
+    @html = @snapshot.html
+    @parent_page = Page.find(@snapshot.page_id)
+    @filename = @parent_page.name.gsub(" ","-") + "-" + @snapshot.created_at.to_s.gsub(" ","-") + ".html"
+    send_data(@html,
+          :filename => @filename,
+          :type => 'text/html',
+          :disposition => 'attachment') 
+  end
```
Comment (GabeIsman) [inline app/controllers/changes_controller.rb:52]: Basically the same notes as the html method, but one additional thing is that you don't need to make these instance variables (prepending the `@`), since you're just using them locally in this method. You might also consider moving the filename logic to the PageSnapshot model. Feels like a more natural home for it.

File: app/controllers/changes_controller.rb (hunk 8 above)
Comment (GabeIsman) [reply app/controllers/changes_controller.rb:28]: It actually occurs to me now that I think all of this logic should be on the PageSnapshotController (which doesn't exist currently). What you're doing doesn't actually have anything to do with changes, it's all based on the snapshots. I don't think it's absolutely necessary to move it, but that would be the better organization. This maybe some of the source of the confusion over what the `change_id` refers to.

File: app/controllers/changes_controller.rb [hunk 12]
```diff
@@ -32,4 +60,5 @@ def set_change
   def change_params
     params.require(:change).permit(:summary)
   end
+
```
Comment (GabeIsman) [inline app/controllers/changes_controller.rb:63]: rm

File: app/views/changes/html.html.erb [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  <div class="container klax-spacer-md">
+    <div class="row klax-spacer-md">
+
+<div class="col-md-12" style="border:solid;"><iframe src="/changes/page/<%= @change_id %>/html/raw"></div> </div>
+      </div>
+
+    </div></div></div>
+    </div>
+  </div><!-- /.container -->
+</div>
```
Comment (GabeIsman) [inline app/views/changes/html.html.erb:39]: Some funky indentation here that makes it hard to parse the HTML structure. Make sure each `<div> </div>` pair is either on the same line as its counterpart or each is on its own line, then reindent the file.

File: app/views/changes/html.html.erb [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+      
+      </div>
+    </div>
+  </div></div>
+<div class="klax-section klax-spacer-lg">
+  <div class="container klax-spacer-md">
+    <div class="row klax-spacer-md">
+
+<div class="col-md-12" style="border:solid;"><iframe src="/changes/page/<%= @change_id %>/html/raw"></div> </div>
```
Comment (GabeIsman) [inline app/views/changes/html.html.erb:33]: Better to use the rails url help here `raw_html_path(@change_id)`, that way if the url structure changes, this file won't need to.

File: config/initializers/assets.rb [hunk 15]
```diff
@@ -9,3 +9,4 @@
 # Precompile additional assets.
 # application.js, application.css, and all non-JS/CSS in app/assets folder are already added.
 # Rails.application.config.assets.precompile += %w( search.js )
+Rails.application.config.assets.precompile += %w( changes.css )
```
Comment (GabeIsman) [inline config/initializers/assets.rb:12]: This doesn't belong here. To link a stylesheet into the app you should add an import statement at the bottom of `app/assets/stylesheets/application.scss`.

File: config/routes.rb [hunk 16]
```diff
@@ -28,6 +28,9 @@
   end
 
   get '/changes/page/:change_id' => 'changes#page', as: :page_change
+  get '/changes/page/:change_id/html' => 'changes#html', as: :show_html
+  get '/changes/page/:change_id/html/raw' => 'changes#raw_html', as: :raw_html
+  get '/changes/page/:change_id/html/download' => 'changes#download', as: :download_html
```
Comment (GabeIsman) [inline config/routes.rb:33]: Going off the existing pattern here these urls should be
```
/changes/html/:change_id
/changes/raw_html/:change_id
/changes/download_html/:change_id
```
I think also the `as:` helper names should include the name of the model they're associated with so `:show_change_html` for example.

File: DEVELOPING.md (hunk 1 above)
Comment (tommeagher) [reply DEVELOPING.md:20]: Ahhh. That's my mistake. I was introduced to it through Heroku and always assumed they had created foreman.

File: app/assets/stylesheets/changes.scss (hunk 6 above)
Comment (tommeagher) [reply app/assets/stylesheets/changes.scss:8]: Totally fine with that. Wasn't entirely certain how the css for each controller worked. Would this css be limited to just iframes used in views connected to the changes controller? Will move this to the html and it will simplify things.

File: app/controllers/changes_controller.rb (hunk 8 above)
Comment (tommeagher) [reply app/controllers/changes_controller.rb:28]: Ahhh. I was wondering about the Pages controller, but could create a PageSnapshotController instead. Got it.

File: app/controllers/changes_controller.rb (hunk 9 above)
Comment (tommeagher) [reply app/controllers/changes_controller.rb:29]: I'm not aware of this so will have to read up on what ActiveRecord relations are, but I think I generally understand what you're saying here.

File: app/controllers/changes_controller.rb (hunk 12 above)
Comment (tommeagher) [reply app/controllers/changes_controller.rb:63]: ? You want me to delete the new line?

File: app/views/changes/html.html.erb (hunk 13 above)
Comment (tommeagher) [reply app/views/changes/html.html.erb:39]: Oh yeah. Gotcha.

File: app/views/changes/html.html.erb (hunk 14 above)
Comment (tommeagher) [reply app/views/changes/html.html.erb:33]: I was not familiar with that function. Will give it a try.

File: config/initializers/assets.rb (hunk 15 above)
Comment (tommeagher) [reply config/initializers/assets.rb:12]: Got it. Since I'm going to 86 that changes.scss, this should be moot anyway.

File: config/routes.rb (hunk 16 above)
Comment (tommeagher) [reply config/routes.rb:33]: So if I move this to PageSnapshotController, should the route also change? 
It feels like since we're accessing these through the /changes/page/ urls it makes sense to keep them together here. Is that OK?

File: app/controllers/changes_controller.rb (hunk 12 above)
Comment (GabeIsman) [reply app/controllers/changes_controller.rb:63]: Yep! Sorry for the terse shorthand.

File: app/assets/stylesheets/changes.scss (hunk 6 above)
Comment (GabeIsman) [reply app/assets/stylesheets/changes.scss:8]: It's not limited in that way, no. As an organizational convention Rails creates a css file for each controller, but there isn't any automatic scoping or anything.

File: app/controllers/changes_controller.rb (hunk 8 above)
Comment (GabeIsman) [reply app/controllers/changes_controller.rb:28]: Yea, minor mistake I made: controller names are plural, by convention, so it's `PageSnapshotsController`.

File: app/assets/stylesheets/changes.scss (hunk 6 above)
Comment (tommeagher) [reply app/assets/stylesheets/changes.scss:8]: OK. That was my understanding. Definitely didn't want to set a global iframe style with this. Thanks for explaining!

File: app/controllers/changes_controller.rb (hunk 9 above)
Comment (GabeIsman) [reply app/controllers/changes_controller.rb:29]: Don't worry about going too deep, you can just access related data in other tables through methods on the model.

File: app/views/changes/html.html.erb (hunk 14 above)
Comment (GabeIsman) [reply app/views/changes/html.html.erb:33]: They are dynamically created url generation helper methods based off of whatever you name your routes in `routes.rb`. More examples above in this same file.

File: config/routes.rb (hunk 16 above)
Comment (GabeIsman) [reply config/routes.rb:33]: The URL should change to match the controller. Maybe:
```
/page_snapshots/:page_snapshot_id
/page_snapshots/:page_snapshot_id/raw
/page_snapshots/:page_snapshot_id/download
```
The way to think about URLs is ultimately as a "resource locator" so they should reflect the resource that they point to. These are pointing to page snapshots in various formats. If you wanted to make the URL subordinate to the page resource that could work too. I could see something like
```
/page/:page_id/snapshots/:snapshot_id
...
```
also making sense. I don't think it makes a ton of sense to link them to changes though. I think of changes as subordinate to snapshots, not the other way around.

**Comment by tommeagher** [comment]
@GabeIsman, I dealt with all of your comments, and this seems to be working. 

The one remaining question I have is if after creating `page_snapshots_controller.rb`, should the `snapshots` and `setup_compare` methods move from `page_controller.rb` into `page_snapshots_controller.rb`? And then would we move `snapshots.html.erb` from the pages views to the `page_snapshots views`?

Found myself in need of this another half-dozen times last week. Would love to get this merged once you approve.

File: app/models/page_snapshot.rb [hunk 17]
```diff
@@ -43,4 +43,9 @@ def parent
   def blank_match_text?
     self.match_text.blank?
   end
+
+  def filename
+    parent_page = Page.find(self.page_id)
```
Comment (GabeIsman) [inline app/models/page_snapshot.rb:48]: This is better referenced with just `self.page`.

Comment (tommeagher) [reply app/models/page_snapshot.rb]: Good call. Pushed another commit to simplify this.

**Comment by tommeagher** [comment]
Let me see if I can carve out a little time tomorrow to take a stab at that. It makes sense to try to do it now if we can. If I can’t get time in the next couple days then maybe we defer it to another PR. Stand by.

**Comment by tommeagher** [comment]
@GabeIsman, this week has obviously gone straight out the window. I opened #345 to capture this new task of tidying up the page_controller, which I'm happy to do later. If you think it's all right, why don't we merge this now and I'll open another PR for that when I can?

**Comment by tommeagher** [comment]
Thanks for merging, @GabeIsman!
